torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -440,6 +473,282 @@ int BestEffortReplay::findFirstMismatchedID(
+
+// Maps that track information relevant to best effort replay about broadcast
+// axes in consumer that are not in producer
+//
+// For example if we have consumer: T0[i0, b1, b2, i3] and producer:
+// T1[i0, i3]
+//
+// If consumer transformations are:
+// -> T[i0, b1o, b1i, b2o, b2i, i3]
+// -> T[i0*b1i*b2o, b1o, b2i, i3]
+// -> T[i0*b1i*b2o*i3, b1o, b2i]
+//
+// forwarding_map would forward i0->i0*b1i*b2o
+// compliment_map would have the entry i0*b1i->b1o and
+// i0*b1i*b2o -> b2i
+//
+// The first is to fast forward transformations in consumer involving broadcast
+// axes not in producer. The compliment map is to track what leaf nodes we may
+// have after we do this. Leaf nodes are only important for replayCasP. Forward
+// map is used for replayCasP and replayPasC.
+struct ConsumerForwardingInfo {
+ public:
+  // Map IterDomain* axes that can safely be forwarded to their output.
+  std::unordered_map<IterDomain*, IterDomain*> forwarding_map;
+
+  // Given a forward id map id_input -> id_forwarded
+  // Track the other inputs in the expr that id_input is an input to. These will
+  // be used to adjust the replay's leaf tracking. Don't need to track one to
+  // many as currently transformations on IterDomains can only have maximum 2
+  // inputs, but maybe in the future we'll have more.
+  std::unordered_map<IterDomain*, std::vector<IterDomain*>> compliment_map;
+
+  ConsumerForwardingInfo(
+      const TensorView* producer,
+      const TensorView* consumer) {
+    // Collect which root axes are in consumer that are not in producer because
+    // of broadcasting
+    std::unordered_set<IterDomain*> consumer_bcast_roots_not_in_producer;
+
+    const auto c2p_root_map =
+        PairwiseRootDomainMap(producer, consumer)
+            .mapConsumerToProducer(consumer->domain(), producer->domain());
+
+    for (auto consumer_root_id : consumer->getRootDomain()) {
+      if (consumer_root_id->isBroadcast()) {
+        if (c2p_root_map.find(consumer_root_id) == c2p_root_map.end()) {
+          consumer_bcast_roots_not_in_producer.emplace(consumer_root_id);
+        }
+      }
+    }
+
+    // We have root axes in consumer that don't exist in producer, now forward
+    // those to include all id's in consumer comprised of only axes not in
+    // producer.
+    auto consumer_bcast_ids_not_in_producer =
+        consumer_bcast_roots_not_in_producer;
+
+    std::vector<Expr*> consumer_history = ExprSort::getExprs(
+        FusionGuard::getCurFusion(),
+        std::vector<Val*>(
+            consumer->domain()->domain().begin(),
+            consumer->domain()->domain().end()));
+
+    auto isValInConsumer =
+        [&consumer_bcast_ids_not_in_producer](Val* input_val) {
+          if (input_val->isA<IterDomain>()) {
+            return consumer_bcast_ids_not_in_producer.find(
+                       input_val->as<IterDomain>()) !=
+                consumer_bcast_ids_not_in_producer.end();
+          } else {
+            return true;
```

Comment:
Why this returns true when `input_val` is not an `IterDomain`?

Reply:
It's just for the all_of check below, but I should simply filter the inputs so they're iter domains and ignore the others. Good point.

Reply:
Updated now.